aten/src/ATen/TensorBase.h

```diff
@@ -2,107 +2,52 @@
+    if (tensor_impl == nullptr) {
       throw std::runtime_error("TensorBaseImpl with nullptr not supported");
     }
-    if(should_retain && pImpl != UndefinedTensor::singleton()) {
-      retain();
+    if (retain) {
+      tensor_impl->_raw_incref();
     }
   }
-  TensorBaseImpl(const TensorBaseImpl & rhs)
-  : pImpl(rhs.pImpl) {
-    if (pImpl != UndefinedTensor::singleton()) {
-      retain();
-    }
-  }
-  TensorBaseImpl(TensorBaseImpl && rhs) noexcept
-  : pImpl(rhs.pImpl) {
-    rhs.pImpl = UndefinedTensor::singleton();
-  }
-  ~TensorBaseImpl() {
-    if (pImpl != UndefinedTensor::singleton()) {
-      release();
-    }
-  }
-  TensorBaseImpl & operator=(TensorBaseImpl && rhs) & {
-    rhs.swap(*this);
-    return *this;
-  }
-  TensorBaseImpl & operator=(TensorBaseImpl const & rhs) & {
-    //TensorBaseImpl ctor retains original rhs.pImpl
-    //then rhs.pImpl is swapped with this->pImpl
-    //finally TensorBaseImpl dtor releases rhs.pImpl, which was originally this->pImpl
-    TensorBaseImpl(rhs).swap(*this);
-    return *this;
-  }
+  TensorBase(c10::intrusive_ptr<TensorImpl, UndefinedTensor>&& ptr) : tensor_impl_(std::move(ptr)) {}
+  TensorBase(const c10::intrusive_ptr<TensorImpl, UndefinedTensor>& ptr) : tensor_impl_(ptr) {}
+
   int64_t dim() const {
-    if (is_strong) {
-      return pImpl->dim();
-    } else {
-      AT_ERROR("Can't call dim() on a WeakTensor");
-    }
-  }
-  void reset() {
-    TensorBaseImpl().swap(*this);
+    return tensor_impl_->dim();
   }
-  void reset(TensorImpl * rhs) {
-    TensorBaseImpl(rhs, true).swap(*this);
-  }
-  void reset(TensorImpl * rhs, bool should_retain) {
-    TensorBaseImpl(rhs, should_retain).swap(*this );
-  }
-  void swap(TensorBaseImpl & rhs) {
-    TensorImpl * tmp = pImpl;
-    pImpl = rhs.pImpl;
-    rhs.pImpl = tmp;
+
+  TensorImpl * unsafeGetTensorImpl() const {
+    return tensor_impl_.get();
   }
-  TensorImpl * get() const {
-    return pImpl;
+  TensorImpl * unsafeReleaseTensorImpl() {
+    return tensor_impl_.release();
   }
-  TensorImpl * detach() {
-    TensorImpl * ret = pImpl;
-    pImpl = UndefinedTensor::singleton();
-    return ret;
+  const c10::intrusive_ptr<TensorImpl, UndefinedTensor>& getIntrusivePtr() const {
+    return tensor_impl_;
   }
 
   bool defined() const {
-    return pImpl != UndefinedTensor::singleton();
+    return tensor_impl_;
   }
 
-  friend struct at::Type;
-
-  //TODO(zach): sort out friend structes
-public:
-  TensorImpl * pImpl;
-
-private:
-  void retain() {
-    if (is_strong) {
-      pImpl->retain();
-    } else {
-      pImpl->weak_retain();
-    }
+  void reset() {
+    tensor_impl_.reset();
   }
 
-  void release() {
-    if (is_strong) {
-      pImpl->release();
-    } else {
-      pImpl->weak_release();
-    }
-  }
-};
+  friend struct WeakTensor;
 
-using TensorBase = TensorBaseImpl<true>;
-using WeakTensorBase = TensorBaseImpl<false>;
+protected:
+  c10::intrusive_ptr<TensorImpl, UndefinedTensor> tensor_impl_;
```

Comment:
Two reasons: (1) Storage prior to this patch was using `storage_impl_` so I wanted to make it consistent, and (2) pImpl was historically used to refer to `TensorImpl*` raw pointer, so when I changed the type I wanted force myself to update all of the call sites (in case something typechecked, but was wrong).